### PR TITLE
Moving compute_nodal_gradient_process to a cpp

### PR DESCRIPTION
**Pull request by loumalouomega**
Some cleaning in fast_trasnfer_between_model_parts_process

### CODE REVIEW

**Comment by loumalouomega** [comment]
Ping

**Comment by loumalouomega** [comment]
I added a cpp test, I found some problems with the MMG so I am debugging step by step

**Comment by loumalouomega** [comment]
Ping

File: kratos/processes/compute_nodal_gradient_process.cpp [hunk 1]
```diff
@@ -0,0 +1,590 @@
...
+ComputeNodalGradientProcess<2, Variable<double>, Historical>::ComputeNodalGradientProcess(
+    ModelPart& rModelPart, 
+    Variable<double>& rOriginVariable, 
+    Variable<array_1d<double,3> >& rGradientVariable, 
+    Variable<double>& rAreaVariable)
+    :mrModelPart(rModelPart), mrOriginVariable(rOriginVariable), mrGradientVariable(rGradientVariable), mrAreaVariable(rAreaVariable)
+{
+    KRATOS_TRY
+    
+    KRATOS_ERROR_IF_NOT(rModelPart.GetNodalSolutionStepVariablesList().Has( rOriginVariable )) << "Missing variable " << rOriginVariable;
```
Comment (rubenzorrilla) [inline kratos/processes/compute_nodal_gradient_process.cpp]: You can use
~~~cpp
    const auto& r_node = *rModelPart.NodesBegin();
    KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(rOriginVariable, r_node);
~~~

File: kratos/processes/compute_nodal_gradient_process.cpp [hunk 2]
```diff
@@ -0,0 +1,590 @@
...
+    KRATOS_TRY;
+    
+    // Set to zero
+    ClearGradient();
+    
+    bounded_matrix<double,TDim+1,  TDim> DN_DX;
+    array_1d<double,TDim+1> N;
+    double Volume;
+    
+    #pragma omp parallel for private(DN_DX,  N,  Volume)
```
Comment (rubenzorrilla) [inline kratos/processes/compute_nodal_gradient_process.cpp:41]: Do you really need the private statement? Why you don't declare DN_DX, N and Volume inside the parallel region?

File: kratos/processes/compute_nodal_gradient_process.cpp [hunk 3]
```diff
@@ -0,0 +1,590 @@
...
+ComputeNodalGradientProcess<2, Variable<double>, NonHistorical>::ComputeNodalGradientProcess(
+    ModelPart& rModelPart, 
+    Variable<double>& rOriginVariable, 
+    Variable<array_1d<double,3> >& rGradientVariable, 
+    Variable<double>& rAreaVariable)
+    :mrModelPart(rModelPart), mrOriginVariable(rOriginVariable), mrGradientVariable(rGradientVariable), mrAreaVariable(rAreaVariable)
+{
+    KRATOS_TRY
+    
+    KRATOS_ERROR_IF_NOT(rModelPart.GetNodalSolutionStepVariablesList().Has( rOriginVariable )) << "Missing variable " << rOriginVariable;
```
Comment (rubenzorrilla) [inline kratos/processes/compute_nodal_gradient_process.cpp]: Same as above.

File: kratos/processes/compute_nodal_gradient_process.cpp [hunk 4]
```diff
@@ -0,0 +1,590 @@
...
+ComputeNodalGradientProcess<3, Variable<double>, Historical>::ComputeNodalGradientProcess(
+    ModelPart& rModelPart, 
+    Variable<double>& rOriginVariable, 
+    Variable<array_1d<double,3> >& rGradientVariable, 
+    Variable<double>& rAreaVariable)
+    :mrModelPart(rModelPart), mrOriginVariable(rOriginVariable), mrGradientVariable(rGradientVariable), mrAreaVariable(rAreaVariable)
+{
+    KRATOS_TRY
+    
+    KRATOS_ERROR_IF_NOT(rModelPart.GetNodalSolutionStepVariablesList().Has( rOriginVariable )) << "Missing variable " << rOriginVariable;
```
Comment (rubenzorrilla) [inline kratos/processes/compute_nodal_gradient_process.cpp]: Same.

File: kratos/processes/compute_nodal_gradient_process.cpp [hunk 5]
```diff
@@ -0,0 +1,590 @@
...
+ComputeNodalGradientProcess<3, Variable<double>, NonHistorical>::ComputeNodalGradientProcess(
+    ModelPart& rModelPart, 
+    Variable<double>& rOriginVariable, 
+    Variable<array_1d<double,3> >& rGradientVariable, 
+    Variable<double>& rAreaVariable)
+    :mrModelPart(rModelPart), mrOriginVariable(rOriginVariable), mrGradientVariable(rGradientVariable), mrAreaVariable(rAreaVariable)
+{
+    KRATOS_TRY
+    
+    KRATOS_ERROR_IF_NOT(rModelPart.GetNodalSolutionStepVariablesList().Has( rOriginVariable )) << "Missing variable " << rOriginVariable;
```
Comment (rubenzorrilla) [inline kratos/processes/compute_nodal_gradient_process.cpp]: :top:

File: kratos/processes/compute_nodal_gradient_process.cpp [hunk 6]
```diff
@@ -0,0 +1,590 @@
...
+ComputeNodalGradientProcess<2, component_type, Historical>::ComputeNodalGradientProcess(
+    ModelPart& rModelPart, 
+    component_type& rOriginVariable, 
+    Variable<array_1d<double,3> >& rGradientVariable, 
+    Variable<double>& rAreaVariable)
+    :mrModelPart(rModelPart), mrOriginVariable(rOriginVariable), mrGradientVariable(rGradientVariable), mrAreaVariable(rAreaVariable)
+{
+    KRATOS_TRY
+    
+    KRATOS_ERROR_IF_NOT(rModelPart.GetNodalSolutionStepVariablesList().Has( rOriginVariable )) << "Missing variable " << rOriginVariable;
```
Comment (rubenzorrilla) [inline kratos/processes/compute_nodal_gradient_process.cpp]: :top:

File: kratos/processes/compute_nodal_gradient_process.cpp [hunk 7]
```diff
@@ -0,0 +1,590 @@
...
+ComputeNodalGradientProcess<2, component_type, NonHistorical>::ComputeNodalGradientProcess(
+    ModelPart& rModelPart, 
+    component_type& rOriginVariable, 
+    Variable<array_1d<double,3> >& rGradientVariable, 
+    Variable<double>& rAreaVariable)
+    :mrModelPart(rModelPart), mrOriginVariable(rOriginVariable), mrGradientVariable(rGradientVariable), mrAreaVariable(rAreaVariable)
+{
+    KRATOS_TRY
+    
+    KRATOS_ERROR_IF_NOT(rModelPart.GetNodalSolutionStepVariablesList().Has( rOriginVariable )) << "Missing variable " << rOriginVariable;
```
Comment (rubenzorrilla) [inline kratos/processes/compute_nodal_gradient_process.cpp]: :top:

File: kratos/processes/compute_nodal_gradient_process.cpp [hunk 8]
```diff
@@ -0,0 +1,590 @@
...
+ComputeNodalGradientProcess<3, component_type, Historical>::ComputeNodalGradientProcess(
+    ModelPart& rModelPart, 
+    component_type& rOriginVariable, 
+    Variable<array_1d<double,3> >& rGradientVariable, 
+    Variable<double>& rAreaVariable)
+    :mrModelPart(rModelPart), mrOriginVariable(rOriginVariable), mrGradientVariable(rGradientVariable), mrAreaVariable(rAreaVariable)
+{
+    KRATOS_TRY
+    
+    KRATOS_ERROR_IF_NOT(rModelPart.GetNodalSolutionStepVariablesList().Has( rOriginVariable )) << "Missing variable " << rOriginVariable;
```
Comment (rubenzorrilla) [inline kratos/processes/compute_nodal_gradient_process.cpp]: :top:

File: kratos/processes/compute_nodal_gradient_process.cpp [hunk 9]
```diff
@@ -0,0 +1,590 @@
...
+ComputeNodalGradientProcess<3, component_type, NonHistorical>::ComputeNodalGradientProcess(
+    ModelPart& rModelPart, 
+    component_type& rOriginVariable, 
+    Variable<array_1d<double,3> >& rGradientVariable, 
+    Variable<double>& rAreaVariable)
+    :mrModelPart(rModelPart), mrOriginVariable(rOriginVariable), mrGradientVariable(rGradientVariable), mrAreaVariable(rAreaVariable)
+{
+    KRATOS_TRY
+    
+    KRATOS_ERROR_IF_NOT(rModelPart.GetNodalSolutionStepVariablesList().Has( rOriginVariable )) << "Missing variable " << rOriginVariable;
```
Comment (rubenzorrilla) [inline kratos/processes/compute_nodal_gradient_process.cpp]: :top:

File: kratos/processes/compute_nodal_gradient_process.h [hunk 10]
```diff
@@ -256,12 +214,12 @@ class ComputeNodalGradientProcess
...
-     * @param geom: The geometry of the element
-     * @param i: The node index
-     * @param k: The component index
+     * @param ThisGeometry The geometry of the element
+     * @param i The node index
+     * @param k The component index
      */
     double& GetGradient(
-        Element::GeometryType& geom,
+        Element::GeometryType& ThisGeometry,
```
Comment (rubenzorrilla) [inline kratos/processes/compute_nodal_gradient_process.h]: Taking into account that this is an aesthetic change, set rThisGeometry as argument name.

File: kratos/tests/processes/test_compute_nodal_gradient_process.cpp [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+                      
+            // Compute gradient
+            typedef ComputeNodalGradientProcess<3, Variable<double>, Historical> GradientType;
+            GradientType process = GradientType(this_model_part, DISTANCE, DISTANCE_GRADIENT);
+            process.Execute();
+            
+//             // DEBUG         
+//             GiDIODebugGradient(this_model_part);
+            
+            const double tolerance = 1.0e-4;
```
Comment (rubenzorrilla) [inline kratos/tests/processes/test_compute_nodal_gradient_process.cpp]: Maybe this tolerance is too large?

File: kratos/processes/compute_nodal_gradient_process.cpp (hunk 2 above)
Comment (loumalouomega) [reply kratos/processes/compute_nodal_gradient_process.cpp:41]: That code is from @RiccardoRossi

File: kratos/tests/processes/test_compute_nodal_gradient_process.cpp (hunk 11 above)
Comment (loumalouomega) [reply kratos/tests/processes/test_compute_nodal_gradient_process.cpp]: Maybe

**Comment by loumalouomega** [comment]
Done

File: kratos/processes/compute_nodal_gradient_process.cpp (hunk 2 above)
Comment (rubenzorrilla) [reply kratos/processes/compute_nodal_gradient_process.cpp:41]: Let's wait for his answer then.

Comment (loumalouomega) [reply kratos/processes/compute_nodal_gradient_process.cpp:41]: I think that is exactly the same, if you use private in creates a copy, that is the same as declare N times inside the parallel region

Comment (rubenzorrilla) [reply kratos/processes/compute_nodal_gradient_process.cpp:41]: Totally agree, but I was wondering if there's another reason to declare it outside the parallel region.

Comment (loumalouomega) [reply kratos/processes/compute_nodal_gradient_process.cpp:41]: Ping @RiccardoRossi

**Review by rubenzorrilla — APPROVED** [review]

**Comment by loumalouomega** [comment]
Merci

**Pull request merged**